### DISCUSSION TITLE
Use inquirer without commander

### DISCUSSION THREAD

**Discussion by stefaanMLB** [jmcdo29/nest-commander#1053] [Q&A] [answered]
Is it possible to use only the inquirer part for nest-commander in a classic REST API ?
I'd like to have a convenient CLI interface on a running application to sent commands to the application.
I know it can also be done by creating extra rest and-points and leverage tools like postman, but a direct CLI interface seems more convenient in production (and even during development and testing)

**Comment by jmcdo29** [comment]
I'm not entirely sure what sense this makes. Inquirer is for asking questions and getting responses through the terminal. How would that work in your REST API?

**Reply by stefaanMLB** [reply]
Mostly for house-keeping tasks like
- query/change/reload configuration without having to restart the application
- trigger scheduled tasks outside of the set schedule
- query global application state (f.i. usage counters)
- manually testing some functions that are also available in the API endpoints

As a -trivial - example, all my backends expose a version endpoint that provides that application name, version, environment, commit ID and build time.  When working on a production docker machine, it would be more convenient to `docker attach` to the running container to get the version details than to open postman and create an API call.

**Reply by smolinari** [reply]
You'd still need to have some way to run the commands that do these tasks, wouldn't you?

Scott

**Reply by stefaanMLB** [reply]
I had hoped that that would not be necessary.
In my mind the inquirer runs async parallel to the rest api and can be used from the command-line at any time.  If needed, the interaction could start with a carriage-return after which the first level of the command-tree is printed and the user can interact until some action is executed on one of the providers of the nest application.

**Reply by jmcdo29** [reply]
The CLI is a **separate** application from the running server. You'd want something closer to Nest's [REPL](https://docs.nestjs.com/recipes/repl)

**Reply by stefaanMLB** [reply]
Hmm, Nest's REPL comes close but is too general and unguided.  
I'd prefer that ops people are guided with a question/answer interaction and to expose only the functions that I chose, even if it means writing up a bunch of questions.
Inquirer is ideal, but I need to be able to run it in parallel with the `app.listen()` task.

**Comment by stefaanMLB** [comment]
It works with [`inquirer`](https://www.npmjs.com/package/inquirer) like below, but `inquirer` doesn't seem to play nicely with typescript, I need a `@ts-ignore` to make it work.
Having the decorator syntax of nest-commander available would be great

```ts
import { NestFactory } from '@nestjs/core';
import { AppModule } from './app.module';
import * as inquirer from 'inquirer';
import { questions } from './pizza';

async function bootstrap() {
  const app = await NestFactory.create(AppModule);
  app.listen(3000);
  while (true) {
    //@ts-ignore
    const answer = await inquirer.prompt(questions);
    console.log(answer);
    // call a nest service with answer
  }
}
bootstrap();

```

```bash
[Nest] 22940  - 02/10/2023 17:14:24     LOG [NestFactory] Starting Nest application...
[Nest] 22940  - 02/10/2023 17:14:24     LOG [InstanceLoader] AppModule dependencies initialized +12ms
? Is this for delivery? (y/N) [Nest] 22940  - 02/10/2023 17:14:24     LOG [RoutesResolver] AppController {/}: +50ms
[Nest] 22940  - 02/10/2023 17:14:24     LOG [RouterExplorer] Mapped {/, GET} route +5ms
[Nest] 22940  - 02/10/2023 17:14:24     LOG [NestApplication] Nest application successfully started +3ms
? Is this for delivery? Yes
? What's your phone number? 0123546879
? What size do you need? Large
? How many do you need? 2
? What about the toppings? Pepperoni and cheese
? You also get a free 2L beverage 7up
? Any comments on your purchase experience? Nope, all good!
{
  toBeDelivered: true,
  phone: '0123546879',
  size: 'large',      
  quantity: 2,
  toppings: 'PepperoniCheese',
  beverage: '7up',
  comments: 'Nope, all good!'
}
? Is this for delivery? (y/N)
```

**Reply by jmcdo29** [reply, marked as answer]
So, _technically_, you can use `imports: [CommandRunnerModule.forRoot()]` to create the module that provides and exports the `InquirerService`. It also provides the `CommandRunnerService`, but that should be okay, it'll just look for `@Command()` classes, find nothing, and set nothing up from them.

**Reply by stefaanMLB** [reply]
There doesn't seem to be a `forRoot()` method on `CommandRunnerModule`, did you mean `forModule()` ?

I tried this `main.ts`

```ts
async function bootstrap() {
  const app = await NestFactory.create(AppModule);
  app
    .get(InquirerService)
    .ask('personInfo', {})
    .then((anser) => console.log(anser));
  await app.listen(3000);
}
bootstrap();
```

with this `app.module`
```ts
@Module({
  imports: [CommandRunnerModule.forModule()],
  controllers: [AppController],
  providers: [AppService],
})
export class AppModule {}
```

and this question
```ts
@QuestionSet({ name: 'personInfo' })
export class PersonInfoQuestions {
  @Question({
    type: 'input',
    name: 'personName',
    message: 'What is your name?',
  })
  parseName(val: string) {
    return val;
  }

  @Question({
    type: 'input',
    name: 'age',
    message: 'How old are you?',
  })
  parseAge(val: string) {
    return Number.parseInt(val, 10);
  }
}
```

But there seems to be a `discoveredClass` missing
```log
TypeError: Cannot read properties of undefined (reading 'discoveredClass')
    at C:\AMD\p-proj\nest-commander-test\packages\nest-commander\src\inquirer.service.ts:66:39
    at C:\AMD\p-proj\nest-commander-test\node_modules\@golevelup\nestjs-discovery\src\discovery.service.ts:74:59
    at Array.filter (<anonymous>)
    at DiscoveryService.providers (C:\AMD\p-proj\nest-commander-test\node_modules\@golevelup\nestjs-discovery\src\discovery.service.ts:74:45)       
    at DiscoveryService.providerMethodsWithMetaAtKey (C:\AMD\p-proj\nest-commander-test\node_modules\@golevelup\nestjs-discovery\src\discovery.service.ts:187:23)
    at InquirerService.findQuestionSet (C:\AMD\p-proj\nest-commander-test\packages\nest-commander\src\inquirer.service.ts:64:7)
    at InquirerService.prompt (C:\AMD\p-proj\nest-commander-test\packages\nest-commander\src\inquirer.service.ts:49:26)
```

**Reply by jmcdo29** [reply]
> There doesn't seem to be a forRoot() method on CommandRunnerModule, did you mean forModule() ?

yep, whoops. Muscle memory, `forRoot` is much more commonly used, this isn't a usual dynamic module so I went with a different naming scheme to kind of separate the idea in my head.

> But there seems to be a discoveredClass missing

That would mean that the InquirerService did not find the class with the `personInfo` question set metadata. Did you add the `PersonInfoQuestion` to the `providers` array?

**Reply by stefaanMLB** [reply]
Yes, That's it, now it works.
Thanks a LoOoOoOoOoT !


Here's the new `app.module.ts` for reference

```ts
import { Module } from '@nestjs/common';
import { AppController } from './app.controller';
import { AppService } from './app.service';
import { CommandRunnerModule } from 'nest-commander';
import { PersonInfoQuestions } from './inquirer/person-info-questions';

@Module({
  imports: [CommandRunnerModule.forModule()],
  controllers: [AppController],
  providers: [AppService, PersonInfoQuestions],
})
export class AppModule {}
```

This is probably the better `main.ts` version
```ts
async function bootstrap() {
  const app = await NestFactory.create(AppModule);
  app.listen(3000);
  while (true) {
    const answer = await app.get(InquirerService).ask('personInfo', {});
    console.log(answer);
  }
}
bootstrap();
```
